config/rbac/service_account.yaml

```diff
@@ -10,57 +10,3 @@ metadata:
     app.kubernetes.io/managed-by: kustomize
   name: controller-manager
   namespace: system
----
-apiVersion: v1
-kind: ServiceAccount
-metadata:
-  name: openstackprovisionserver
----
-apiVersion: rbac.authorization.k8s.io/v1
-kind: Role
-metadata:
-  name: openstackprovisionserver-role
-  namespace: openstack
-rules:
-- apiGroups:
-  - security.openshift.io
-  resourceNames:
-  - privileged
-  resources:
-  - securitycontextconstraints
-  verbs:
-  - use
-- apiGroups:
-  - baremetal.openstack.org
-  resources:
-  - openstackprovisionservers
-  verbs:
-  - get
-  - list
-- apiGroups:
-  - baremetal.openstack.org
-  resources:
-  - openstackprovisionservers/status
-  verbs:
-  - get
-  - list
-  - update
```

Comment:
The service account will still need these permissions, I believe, for the provisioning agent to be able to update the `OpenStackProvisionServer` status to set the discovered provisioning IP

Reply:
I added the missing ones to rbac rules. Hope that would be enough.

Reply:
I am seeing this when I build the operator and deploy it, and then `oc create -f` an `OSProvServer`:

```
2023-06-02T12:54:35Z	ERROR	Reconciler error	{"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"name":"openstack","namespace":"openstack"}, "namespace": "openstack", "name": "openstack", "reconcileID": "b70977e3-bca2-4db4-9aa4-d848ac92db86", "error": "Error creating role provisionserver-openstack-role *v1.Role openstack/provisionserver-openstack-role: roles.rbac.authorization.k8s.io \"provisionserver-openstack-role\" is forbidden: user \"system:serviceaccount:openstack-operators:openstack-baremetal-operator-controller-manager\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:openstack-operators\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held:\n{APIGroups:[\"baremetal.openstack.org\"], Resources:[\"openstackprovisionservers/status\"], Verbs:[\"list\"]}"}
```

Reply:
Probably because https://github.com/openstack-k8s-operators/openstack-baremetal-operator/blob/main/controllers/openstackprovisionserver_controller.go#L70 have list? updated the patch.

Reply:
Thanks for testing it. I've not been able to test it yet :smile:

Reply:
I see this on the `Deployment` for the provision server:

```
  - lastTransitionTime: "2023-06-02T14:52:08Z"
    lastUpdateTime: "2023-06-02T14:52:08Z"
    message: 'pods "openstack-openstackprovisionserver-5fb58d4cbc-" is forbidden:
      unable to validate against any security context constraint: [provider anyuid:
      .spec.securityContext.hostNetwork: Invalid value: true: Host network is not
      allowed to be used, spec.initContainers[0].securityContext.hostNetwork: Invalid
      value: true: Host network is not allowed to be used, spec.containers[0].securityContext.hostNetwork:
      Invalid value: true: Host network is not allowed to be used, spec.containers[1].securityContext.hostNetwork:
      Invalid value: true: Host network is not allowed to be used, provider restricted-v2:
      .spec.securityContext.hostNetwork: Invalid value: true: Host network is not
      allowed to be used, spec.containers[0].securityContext.runAsUser: Invalid value:
      0: must be in the ranges: [1000750000, 1000759999], provider "restricted": Forbidden:
      not usable by user or serviceaccount, provider "containerized-data-importer":
      Forbidden: not usable by user or serviceaccount, provider "nonroot-v2": Forbidden:
      not usable by user or serviceaccount, provider "nonroot": Forbidden: not usable
      by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable
      by user or serviceaccount, provider "kubevirt-controller": Forbidden: not usable
      by user or serviceaccount, provider "machine-api-termination-handler": Forbidden:
      not usable by user or serviceaccount, provider "bridge-marker": Forbidden: not
      usable by user or serviceaccount, provider "hostnetwork-v2": Forbidden: not
      usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable
      by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user
      or serviceaccount, provider "linux-bridge": Forbidden: not usable by user or
      serviceaccount, provider "kubevirt-handler": Forbidden: not usable by user or
      serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount,
      provider "privileged": Forbidden: not usable by user or serviceaccount]'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
```

I will tinker with the permissions this afternoon to see what is missing.

Reply:
Ah, I know what the problem is here.  This SA/Role is a special role created for the `OpenStackProvisionServer` CR instances.  Just like the new SA, Roles and RoleBindings that are now dynamically created, this also needs to be created within the reconcile loop for a given `OpenStackProvisionServer` CR within a given namespace.  So it's proper to remove this hardcoded definition as we're doing here, but we need to then represent it in the code (create it) and then use it for the provision server's `Deployment` (instead of `instance.RbacResourceName()`).

Reply:
Probably it's because we use hostNetwork: true for the deployment here, we need to provide more scc permissions, anyuid only won't work. I'll try that as I've seen the same issue as you noticed. 

https://github.com/openstack-k8s-operators/openstack-baremetal-operator/blob/main/pkg/openstackprovisionserver/deployment.go#L99

Reply:
I was able to take your patch and add these two changes.  Then everything seemed to work:

https://github.com/abays/openstack-baremetal-operator/blob/38f25ad1f64e9c83d808fdb85fa14a3e5011e996/controllers/openstackprovisionserver_controller.go#L77

https://github.com/abays/openstack-baremetal-operator/blob/38f25ad1f64e9c83d808fdb85fa14a3e5011e996/controllers/openstackprovisionserver_controller.go#L209-L214

Reply:
Right, I was thinking of making those changes but we probably should use `hostnetwork` rather than `privileged`. Let me try that.